pkg/ccl/streamingccl/streamingest/ingest_span_configs.go

```diff
@@ -0,0 +1,253 @@
+		client:        client,
+		rekeyer:       rekeyer,
+	}, nil
+}
+
+func (sc *spanConfigIngestor) close(ctx context.Context) {
+	if err := sc.client.Close(ctx); err != nil {
+		log.Errorf(ctx, "error on client.close(): %s", err)
+	}
+	if err := sc.group.Wait(); err != nil {
+		log.Errorf(ctx, "error on close(): %s", err)
+	}
+}
+
+func (sc *spanConfigIngestor) ingestSpanConfigs(
+	ctx context.Context, tenantName roachpb.TenantName,
+) error {
+	sub, err := sc.client.SetupSpanConfigsStream(ctx, tenantName)
+	if err != nil {
+		return err
+	}
+
+	g := ctxgroup.WithContext(ctx)
+	g.GoCtx(sub.Subscribe)
+	g.GoCtx(sc.checkForCutover)
+
+	return sc.consumeSpanConfigs(ctx, sub)
+}
+
+func (sc *spanConfigIngestor) checkForCutover(ctx context.Context) error {
+	tick := time.NewTicker(cutoverSignalPollInterval.Get(&sc.settings.SV))
+	defer tick.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-tick.C:
+			cutoverReached, err := sc.cutoverSignal.cutoverReached(ctx)
+			if err != nil {
+				return err
+			}
+			if cutoverReached {
+				close(sc.cutoverCh)
+				return nil
+			}
+		}
+	}
+}
+
+func (sc *spanConfigIngestor) consumeSpanConfigs(
+	ctx context.Context, subscription streamclient.Subscription,
+) error {
+	for {
+		select {
+		case event, ok := <-subscription.Events():
+			if !ok {
+				return nil
+			}
+			if err := sc.consumeEvent(ctx, event); err != nil {
+				return err
+			}
+		case <-sc.cutoverCh:
+			return nil
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+}
+
+func (sc *spanConfigIngestor) consumeEvent(ctx context.Context, event streamingccl.Event) error {
+	switch event.Type() {
+	case streamingccl.SpanConfigEvent:
+		return sc.bufferRecord(ctx, event.GetSpanConfigEvent())
+	case streamingccl.CheckpointEvent:
+		return sc.flushEvents(ctx)
+	default:
+		return errors.AssertionFailedf("received non span config update %s", event)
+	}
+}
+
+func (sc *spanConfigIngestor) bufferRecord(
+	ctx context.Context, update *streampb.StreamedSpanConfigEntry,
+) error {
+	sourceSpan := update.SpanConfig.Target.GetSpan()
+	destStartKey, ok, err := sc.rekeyer.RewriteKey(sourceSpan.Key, 0)
+	if err != nil {
+		return err
+	}
+	if !ok {
+		// No need to replicate the span cfgs for ephemeral tables in the app tenant
+		return nil
+	}
+	destEndKey, ok, err := sc.rekeyer.RewriteKey(sourceSpan.EndKey, 0)
+	if err != nil {
+		return err
+	}
+	if !ok {
+		// No need to replicate the span cfgs for ephemeral tables in the app tenant
+		//
+		// TODO(msbutler): This error handling isn't ideal as the span for this span
+		// cfg doesn't necessarily cover only the ephemeral tables. It would be
+		// better to remove this logic out of the rekeyer.
+		return nil
+	}
+	targetSpan := roachpb.Span{Key: destStartKey, EndKey: destEndKey}
+
+	if sc.lastBufferedSourceTimestamp.Less(update.Timestamp) && len(sc.bufferedUpdates) != 0 {
+		// If this event was originally written at a later timestamp than what's in the buffer, flush the buffer.
```

Comment:
Why do we only check the length of bufferedUpdates here and not also the length of bufferedDeletes?